### PR TITLE
io.Source is not closed when used in most common text reading idiom

### PR DESCRIPTION
**Pull request by chick**
- Provide new tools for reading in text
  - from a file
  - from a string file name
  - from a resource file
- text can be read in as
  - a single string with newlines
  - a seq of strings, one string per line
- FileUtils put in its own file (was in `Driver`)
  - in same package to keep existing API the same

Hopefully this will protect users from creating resource leaks that
may explain sbt crashing.
Also simpler interface should make code more streamlined, existing
uses of io.Source seem often to be converting between text and seq
unnecessarily.

### CODE REVIEW

File: src/main/scala/firrtl/FileUtils.scala [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    * Instead we try to run the executable itself (with innocuous arguments) and interpret any errors/exceptions
+    *  as an indication that the executable is unavailable.
+    */
+  lazy val isVCSAvailable: Boolean = isCommandAvailable(Seq("vcs",  "-platform"))
+
+  /** Read a text file into a Seq of strings
+    * Closes the file after read to avoid dangling file handles
+    *
+    * @param fileName The file to read
+    * @return
```
Comment (seldridge) [inline src/main/scala/firrtl/FileUtils.scala]: Either provide a message here or remove. The lede of `"Read a text file into a Seq of strings"` is good enough if you want to remove the `@return`.

File: src/main/scala/firrtl/FileUtils.scala [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    val lines = source.getLines()
+    source.close()
+    lines.toSeq
+  }
+
+  /** Read a text file into a Seq of strings
+    * Closes the file after read to avoid dangling file handles
+    *
+    * @param file a java File to be read
+    * @return
```
Comment (seldridge) [inline src/main/scala/firrtl/FileUtils.scala]: Put something here or remove.

File: src/main/scala/firrtl/FileUtils.scala [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    val lines = source.getLines()
+    source.close()
+    lines.toSeq
+  }
+
+  /** Read a text file into a single string
+    * Closes the file after read to avoid dangling file handles
+    *
+    * @param fileName The file to read
+    * @return
```
Comment (seldridge) [inline src/main/scala/firrtl/FileUtils.scala]: Put something here or remove.

File: src/main/scala/firrtl/FileUtils.scala [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    val text = source.mkString
+    source.close()
+    text
+  }
+
+  /** Read a text file into a single string
+    * Closes the file after read to avoid dangling file handles
+    *
+    * @param file a java File to be read
+    * @return
```
Comment (seldridge) [inline src/main/scala/firrtl/FileUtils.scala]: Put something here or remove.

File: src/main/scala/firrtl/FileUtils.scala [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    val text = source.mkString
+    source.close()
+    text
+  }
+
+  /** Read text file into a Seq of strings
+    * Closes the file after read to avoid dangling file handles
+    *
+    * @param resourceName a java File to be read
+    * @return
```
Comment (seldridge) [inline src/main/scala/firrtl/FileUtils.scala]: Put something here or remove.

File: src/main/scala/firrtl/FileUtils.scala [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    val text = io.Source.fromInputStream(inputStream).getLines().toSeq
+    inputStream.close()
+    text
+  }
+
+  /** Read text file into a single string
+    * Closes the file after read to avoid dangling file handles
+    *
+    * @param resourceName a java File to be read
+    * @return
```
Comment (seldridge) [inline src/main/scala/firrtl/FileUtils.scala]: Put something here or remove.

File: src/main/scala/firrtl/FileUtils.scala [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+  }
+
+  /** Read text file into a Seq of strings
+    * Closes the file after read to avoid dangling file handles
+    *
+    * @param resourceName a java File to be read
+    * @return
+    */
+  def getLinesResource(resourceName: String): Seq[String] = {
+    val inputStream = getClass.getResourceAsStream("/annotations/SampleAnnotations.anno")
```
Comment (seldridge) [inline src/main/scala/firrtl/FileUtils.scala]: This seems to be accidentally hard-coded. I think the following is right, though?

```suggestion
    val inputStream = getClass.getResourceAsStream(s"/$resourceName")
```

File: src/main/scala/firrtl/FileUtils.scala [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+  }
+
+  /** Read text file into a single string
+    * Closes the file after read to avoid dangling file handles
+    *
+    * @param resourceName a java File to be read
+    * @return
+    */
+  def getTextResource(resourceName: String): String = {
+    val inputStream = getClass.getResourceAsStream("/annotations/SampleAnnotations.anno")
```
Comment (seldridge) [inline src/main/scala/firrtl/FileUtils.scala]:
```suggestion
    val inputStream = getClass.getResourceAsStream(s"/$resourceName")
```

Comment (chick) [reply src/main/scala/firrtl/FileUtils.scala]: Yikes, good catch.

File: src/main/scala/firrtl/FileUtils.scala (hunk 7 above)
Comment (chick) [reply src/main/scala/firrtl/FileUtils.scala]: Yikes, good catch.

Comment (chick) [reply src/main/scala/firrtl/FileUtils.scala]: But I don't think we should hard code the initial slash. In all cases that I see users have provided the initial slash.

File: src/main/scala/firrtl/FileUtils.scala (hunk 1 above)
Comment (chick) [reply src/main/scala/firrtl/FileUtils.scala]: Going with removing `@return` in all cases.

File: src/main/scala/firrtl/FileUtils.scala (hunk 7 above)
Comment (seldridge) [reply src/main/scala/firrtl/FileUtils.scala]: Leaving off the leading `/` makes sense to me.

**Comment by ucbjrl** [comment]
Unfortunately, this (or the follow-on #1143) is causing [ucb-bar/dsptools](https://github.com/ucb-bar/dsptools) to blow up with:
```
Stream Closed
java.io.IOException: Stream Closed
	at java.io.FileInputStream.readBytes(Native Method)
	at java.io.FileInputStream.read(FileInputStream.java:255)
	at sun.nio.cs.StreamDecoder.readBytes(StreamDecoder.java:284)
	at sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:326)
	at sun.nio.cs.StreamDecoder.read(StreamDecoder.java:178)
	at java.io.InputStreamReader.read(InputStreamReader.java:184)
	at java.io.BufferedReader.fill(BufferedReader.java:161)
	at java.io.BufferedReader.readLine(BufferedReader.java:324)
	at java.io.BufferedReader.readLine(BufferedReader.java:389)
	at scala.io.BufferedSource$BufferedLineIterator.hasNext(BufferedSource.scala:74)
	at scala.collection.Iterator.toStream(Iterator.scala:1415)
	at scala.collection.Iterator.toStream$(Iterator.scala:1414)
	at scala.collection.AbstractIterator.toStream(Iterator.scala:1429)
	at scala.collection.TraversableOnce.toSeq(TraversableOnce.scala:301)
	at scala.collection.TraversableOnce.toSeq$(TraversableOnce.scala:301)
	at scala.collection.AbstractIterator.toSeq(Iterator.scala:1429)
	at firrtl.FileUtils$.getLines(FileUtils.scala:116)
	at firrtl.util.BackendCompilationUtilities.verilogToCpp(BackendCompilationUtilities.scala:125)
	at firrtl.util.BackendCompilationUtilities.verilogToCpp$(BackendCompilationUtilities.scala:99)
	at chisel3.testers.TesterDriver$.verilogToCpp(TesterDriver.scala:12)
	at chisel3.testers.TesterDriver$.execute(TesterDriver.scala:60)
	at chisel3.iotesters.ChiselRunners.runTester(ChiselSpec.scala:17)
	at chisel3.iotesters.ChiselRunners.runTester$(ChiselSpec.scala:16)
	at chisel3.iotesters.ChiselFlatSpec.runTester(ChiselSpec.scala:26)
	at chisel3.iotesters.ChiselRunners.assertTesterPasses(ChiselSpec.scala:20)
	at chisel3.iotesters.ChiselRunners.assertTesterPasses$(ChiselSpec.scala:19)
	at chisel3.iotesters.ChiselFlatSpec.assertTesterPasses(ChiselSpec.scala:26)
	at dsptools.numbers.BlackBoxFloatSpec.$anonfun$new$53(BlackBoxFloat.scala:258)
```

**Comment by seldridge** [comment]
Ugh... @chick: I think your fix for `getLinesResource` needs to be extended for both the `getLines` methods. Assuming this is what it is, I should've caught this in the review.

Maybe some tests for `FileUtils` would be good, too.

**Comment by seldridge** [comment]
Rushed out a fix here: https://github.com/freechipsproject/firrtl/pull/1144

**Pull request merged**